### PR TITLE
[2/2] Allow homeservers to send registration emails | Accepting the verification

### PR DESCRIPTION
**Pull request by anoadragon453**
Fixes: #5751
Fixes: https://github.com/matrix-org/synapse/issues/5928

https://github.com/matrix-org/synapse/pull/5835 allowed Synapse to send registration emails to the user. Now we need to accept them and have it succeed the `m.login.email.identity` registration flow step.

`account_threepid_handler` will also be switched from a `str` in the config file to a dictionary which contains entries for `msisdn` and `email`, each with their own `str`. This will let people use an external server to handle `msisdn` registration and password reset requests, while using Synapse for email-based things.

And the `password_servlet` hack that was introduced in https://github.com/matrix-org/synapse/pull/5377/files#diff-b8464485d36f6f87caee3f4d82524213R189 to distinguish a registration call from a password reset call will be removed.

### CODE REVIEW

File: synapse/handlers/register.py [hunk 1]
```diff
@@ -362,58 +360,6 @@ def appservice_register(self, user_localpart, as_token):
         )
         return user_id
 
-    @defer.inlineCallbacks
```
Comment (richvdh) [inline synapse/handlers/register.py]: please can the bits of this PR which kill off old dead code be factored out to a separate PR?

Comment (anoadragon453) [reply synapse/handlers/register.py]: Of course, sorry about that.

Comment (anoadragon453) [reply synapse/handlers/register.py]: https://github.com/matrix-org/synapse/pull/5963

File: synapse/handlers/register.py [hunk 2]
```diff
@@ -360,6 +362,56 @@ def appservice_register(self, user_localpart, as_token):
         )
         return user_id
 
+    @defer.inlineCallbacks
+    def check_recaptcha(self, ip, private_key, challenge, response):
```
Comment (richvdh) [inline synapse/handlers/register.py]: there's something up with this diff now.

**Comment by anoadragon453** [comment]
Factored out the `m.require_identity_server` flag change to https://github.com/matrix-org/synapse/pull/5968

**Comment by anoadragon453** [comment]
Factored out the `account_threepid_delegates` changed to https://github.com/matrix-org/synapse/pull/5969

**Comment by richvdh** [comment]
taking this out of review until #5969 is done

**Comment by anoadragon453** [comment]
#5969 is now merged.

File: synapse/handlers/auth.py [hunk 3]
```diff
@@ -377,9 +365,7 @@ def _check_auth_dict(self, authdict, clientip, password_servlet=False):
         if checker is not None:
             # XXX: Temporary workaround for having Synapse handle password resets
```
Comment (richvdh) [inline synapse/handlers/auth.py]: dead comment?

File: synapse/push/mailer.py [hunk 4]
```diff
@@ -629,6 +629,27 @@ def format_ts_filter(value, format):
     return time.strftime(format, time.localtime(value / 1000))
 
 
+# XXX: This method and the next could likely be combined in a smarter way
+@staticmethod
```
Comment (richvdh) [inline synapse/push/mailer.py]: staticmethod isn't appropriate for a top-level function?

File: synapse/push/mailer.py [hunk 5]
```diff
@@ -629,6 +629,27 @@ def format_ts_filter(value, format):
...
+def load_jinja2_template(template_dir, template_filename, template_vars):
```
Comment (richvdh) [inline synapse/push/mailer.py]: I think it's confusing to have a `load_jinja2_templates` which just returns the templates, and `load_jinja2_template` which also calls `render`. I'd make a case for hoisting the `render` call to the caller which will save the `template_vars` param, but I guess you could rename the function instead.

File: synapse/push/mailer.py (hunk 4 above)
Comment (anoadragon453) [reply synapse/push/mailer.py]: PyCharm lied to me...

File: synapse/push/mailer.py (hunk 5 above)
Comment (anoadragon453) [reply synapse/push/mailer.py]: I've refactored `load_jinja2_templates` into a single function which can do the work of both.

The entire refactor is contained within https://github.com/matrix-org/synapse/pull/5940/commits/2ee17d70823d5b3c20b362df56d2f71418d3ac56.

File: synapse/push/mailer.py [hunk 6]
```diff
@@ -629,25 +629,50 @@ def format_ts_filter(value, format):
...
-def load_jinja2_templates(config, template_html_name, template_text_name):
-    """Load the jinja2 email templates from disk
+def load_jinja2_templates(
+    template_dir,
+    template_filenames,
+    apply_format_ts_filter=False,
+    apply_mxc_to_http_filter=False,
+    public_baseurl=None,
+):
+    """Loads a jinja2 template with variables to insert
```
Comment (richvdh) [inline synapse/push/mailer.py]: this seems to be incorrect?
